src/Azure/Orleans.ScheduledJobs.AzureStorage/AzureStorageJobShardManager.cs

```diff
@@ -0,0 +1,396 @@
+                    _jobShardCache.TryRemove(blob.Name, out _);
+                    await shard.DisposeAsync();
+                    LogShardOwnershipConflict(_logger, blob.Name, siloAddress);
+                    continue;
+                }
+            }
+            else if (!creatorIsActive || siloAddress.Equals(creator))
+            {
+                // If I am not the creator, only reclaim if the creator is not active anymore,
+                // or if for some reason I am the creator but I don't have the shard in cache (should not happen normally)
+                LogClaimingShard(_logger, blob.Name, siloAddress, creatorIsActive, siloAddress.Equals(creator));
+                var blobClient = _client.GetAppendBlobClient(blob.Name);
+                var metadata = blob.Metadata;
+                shard = new AzureStorageJobShard(blob.Name, shardStartTime, maxDueTime, blobClient, metadata, blob.Properties.ETag);
+                if (!await TryTakeOwnership(shard, metadata, siloAddress, cancellationToken))
+                {
+                    // Someone else took over the shard, dispose and continue
+                    await shard.DisposeAsync();
+                    LogShardOwnershipConflict(_logger, blob.Name, siloAddress);
+                    continue;
+                }
+                await shard.InitializeAsync(cancellationToken);
+                // We don't want to add new jobs to shards that we just took ownership of
+                await shard.MarkAsCompleteAsync(cancellationToken);
+                _jobShardCache[blob.Name] = shard;
+            }
+
+            if (shard != null)
+            {
+                LogShardAssigned(_logger, blob.Name, siloAddress);
+                result.Add(shard);
+            }
+        }
+        
+        LogAssignmentCompleted(_logger, result.Count, siloAddress);
+        return result;
+
+        async Task<bool> TryTakeOwnership(AzureStorageJobShard shard, IDictionary<string, string> metadata, SiloAddress newOwner, CancellationToken ct)
+        {
+            metadata["Owner"] = newOwner.ToParsableString();
+            metadata["MembershipVersion"] = _clusterMembership.CurrentSnapshot.Version.Value.ToString();
+            try
+            {
+                await shard.UpdateBlobMetadata(metadata, ct);
+                LogOwnershipTaken(_logger, shard.Id, newOwner);
+                return true;
+            }
+            catch (RequestFailedException ex)
+            {
+                // Someone else took over the shard
+                LogOwnershipFailed(_logger, ex, shard.Id, newOwner);
+                return false;
+            }
+        }
+    }
+
+    public override async Task<IJobShard> RegisterShard(SiloAddress siloAddress, DateTimeOffset minDueTime, DateTimeOffset maxDueTime, IDictionary<string, string> metadata, bool assignToCreator = true, CancellationToken cancellationToken = default)
```

Comment:
We should add tests to the test kit to ensure that behavior of RegisterShard / AssignShards is consistent across providers.